scriptworker/ed25519.py

```diff
@@ -0,0 +1,183 @@
+    """Convert an ed25519 public key to a base64-encoded string.
+
+    Args:
+        key (Ed25519PublicKey): the key to write to the file.
+
+    Returns:
+        str: the key representation as a str
+
+    """
+    return base64.b64encode(key.public_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PublicFormat.Raw,
+    ), None).decode('utf-8')
+
+
+def verify_ed25519_signature(public_key, contents, signature, message):
+    """Verify that ``signature`` comes from ``public_key`` and ``contents``.
+
+    Args:
+        public_key (Ed25519PublicKey): the key to verify the signature
+        contents (bytes): the contents that was signed
+        signature (bytes): the signature to verify
+        message (str): the error message to raise.
+
+    Raises:
+        ScriptWorkerEd25519Error: on failure
+
+    """
+    try:
+        public_key.verify(signature, contents)
+    except (UnsupportedAlgorithm, InvalidSignature) as exc:
+        raise ScriptWorkerEd25519Error(message % {'exc': str(exc)})
+
+
+# verify_ed25519_signature_cmdln {{{1
+def verify_ed25519_signature_cmdln(args=None, exception=SystemExit):
+    """Verify an ed25519 signature from the command line.
+
+    Args:
+        args (list, optional): the commandline args to parse. If ``None``, use
+            ``sys.argv[1:]``. Defaults to ``None``.
+        exception (Exception, optional): the exception to raise on failure.
+            Defaults to ``SystemExit``.
+
+    """
+    args = args or sys.argv[1:]
+    parser = argparse.ArgumentParser(
+        description="""Verify an ed25519 signature from the command line.
+
+Given a file and its detached signature, verify that it has been signed with
+a valid key. This key can be specified on the command line; otherwise we'll
+default to ``config['ed25519_public_keys']``.""")
+    parser.add_argument('--pubkey', help='path to a base64-encoded ed25519 pubkey, optional')
+    parser.add_argument('file_path')
+    parser.add_argument('sig_path')
+    opts = parser.parse_args(args)
+    log = logging.getLogger('scriptworker')
+    log.setLevel(logging.DEBUG)
+    logging.basicConfig()
+    pubkeys = {}
+    if opts.pubkey:
+        pubkeys['cmdln'] = [read_from_file(opts.pubkey)]
+    pubkeys.update(dict(DEFAULT_CONFIG['ed25519_public_keys']))
+    contents = read_from_file(opts.file_path, file_type='binary')
+    signature = read_from_file(opts.sig_path, file_type='binary')
+    for key_type, seeds in pubkeys.items():
+        for seed in seeds:
+            try:
+                verify_ed25519_signature(
+                    ed25519_public_key_from_string(seed), contents, signature,
+                    "didn't work with {}".format(seed)
+                )
+                log.info("Verified good with {} seed {} !".format(
+                    key_type, seed
+                ))
+                sys.exit(0)
+            except ScriptWorkerEd25519Error:
+                pass
+    else:
```

Comment:
The `else` clause isn't needed here and I think it's slightly clearer if you just fall through to it.

Reply:
Removed.